mrmustard/lab_dev/states/base.py

```diff
@@ -827,6 +833,30 @@ def __init__(
         if representation is not None:
             self._representation = representation
 
+    def auto_shape(self) -> tuple[int, ...]:
+        r"""
+        A pretty enough estimate of the Fock shape of this Ket, define)d as the shape of the Fock
+        array (batch excluded) if it exists, and if it doesn't exist it is computed as the shape
+        that captures at least ``settings.AUTOCUTOFF_PROBABILITY`` of the probability mass of each
+        single-mode marginal (default 99.9%).
+        Note that it overwrites any ``None``s appearing in the ``fock_shape`` attribute and that
+        if a state is batched, the auto_shape is calculated for the first element.
+        """
+        try:  # fock
+            shape = self._representation.array.shape[1:]
+        except AttributeError:  # bargmann
+            if None not in self.fock_shape:  # try the fock_shape if ready
+                return tuple(self.fock_shape)
+            repr = self.representation.conj() & self.representation
+            shape = autoshape_numba(
+                math.asnumpy(repr.A[0]),
+                math.asnumpy(repr.b[0]),
+                math.asnumpy(repr.c[0]),
+            )
```

Comment:
See comment in DM section about subnormalized states.

Reply:
done